rules/doxygen.bzl

```diff
@@ -0,0 +1,343 @@
+# Copyright lowRISC contributors (OpenTitan project).
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+load("//hw/top:defs.bzl", "ALL_TOP_NAMES")
+
+DoxygenCcInputInfo = provider(
+    fields = {
+        "files": "depset of files",
+        "include_paths": "depset of include paths",
+        "strip_from_path": "depset of paths to strip",
+    },
+)
+
+def _merge_doxygen_cc_input_infos(infos):
+    """
+    Given a list of DoxygenCcInputInfo, merge them into one by accumulating of files
+    and include paths.
+
+    Args:
+      infos: List of DoxygenCcInputInfo.
+
+    Returns: a DoxygenCcInputInfo.
+    """
+    return DoxygenCcInputInfo(
+        files = depset(transitive = [getattr(info, "files", depset()) for info in infos]),
+        include_paths = depset(transitive = [getattr(info, "include_paths", depset()) for info in infos]),
+        strip_from_path = depset(transitive = [getattr(info, "strip_from_path", depset()) for info in infos]),
+    )
+
+def doxygen_gather_all_package_cc(name):
+    """
+    Create a `cc_library` targets which depends on all `cc_library` targets
+    previously defined in the same BUILD file.
```

Comment:
Just to check, is this order-dependent within the file? i.e. it only gathers `cc_libraries` defined before it and not after, so it needs to go at the end of the file?

Reply:
I think the "previously" wording is enough, I just wanted to be sure

Thanks